### ISSUE TITLE
Kurento 1020 error on many concurrent webcam activations

### ISSUE THREAD

**Issue by basisbit** [bigbluebutton/bigbluebutton#11099]
**Describe the bug**
Enabling the webcam of for example 30 or 50 users within a short time (for example 60 seconds) causes 1020 errors for many of the clients and thus their webcam can't be shared. If users try again when there aren't many other users trying to enable webcam sharing at the same time, it succeeds.

This causes errors for the typical "30 students from one class switch to another BBB room and try to enable their webcam again", and those errors seem to not be logged on the server.

Looking at the log, it seems like Kurento processes the webcam-connecting workflow one after another instead of concurrently in multiple threads. Might be regression caused by https://github.com/Kurento/kms-core/pull/22 ? (Edit: nope, apparently that is not the reason for the problem, see comments below)

**To Reproduce**
1. have 50 users (or 30 users on a VM running bbb) join a session, 
2. During a few seconds of timespan, for example 30 seconds or one minute, have all the users enable their webcam sharing.
3. Now you get many failure reports.
4. While users experience this problem, try sharing your webcam and see the error also showing up.
5. wait a few minutes and try sharing again << this usually works again.

**Expected behavior**
Webcam sharing should work even if bunch of people turn on their webcam in a BBB session roughly at the same time.

**Additional context**
- Tested on BBB 2.2.30 and BBB 2.2.31 on different setups from different companies.
- Nginx CPU load does not go above 20%
- Kurento load goes up a lot but htop does not show any CPU thread going above 70% load
- TURN/STUN server CPU load does not increase noticeably (clients were connected listenonly already previously -> cache)
- happens with 3-split-Kurento as well as without
- increasing nginx `NOFILE` or `worker_rlimit_nofile` or `worker_connections` does not solve the problem
- nothing unusual in Nginx error logs (but might be caused by log level)
- No error/failed/exception in Kurento debug log

The Kurento logs show that the connecting process was tried/done 1744 times by Kurento even just within those 60 seconds. Referring to the process of this:
```
MediaStateChanged 
-> MediaFlowInStateChange 
-> MediaFlowInStateChange 
-> MediaFlowOutStateChange 
-> OnIceCandidate 
-> IceGatheringDone 
-> NewCandidatePairSelected 
-> "operation":"processOffer","operationParams":{"offer":"v=0\r\no...
```

**Comment by prlanzarin** [comment]
> Looking at the log, it seems like Kurento processes the webcam-connecting workflow one after another instead of concurrently in multiple threads. Might be regression caused by Kurento/kms-core#22 .

I doubt this is a regression from the worker pool rewrite. I was the one that diagnosed that issue, sent data, failure points and reproducible steps on it (along with a much simpler example workaround) to Juan. If anything, we had it **way worse** before the rewrite because 1) it was slower 2) the old worker pool would spin lock into a thread spawning boogaloo which would take the whole server down.

The new rewrite limits the signalling plane workerpool up to the number of virtual cores available. And that worker pool is close to useless because it handles some very specific outbound event-firing thing.
If you do think this is a regression and have a reproducible use case, please try it with Kurento 6.13 (there are docker images available).

AFAIK what you're describing has been happening for eons and is part of a deeper problem of how big the footprint is when fully negotiating a endpoint in Kurento. And how the one peer per media stream approach gets us into an O(n²) complexity hell. And also how poorly the media probes are implemented in KMS (which yes, might have to do with thread starvation). 

The 1020 happens because a client-side endpoint relies on a media flowing event coming from the server to know the thing worker. The larger the number of the endpoints, the slower it is for the media probes to find out that a keyframe has flown through. And hence the slower it is for the event to traverse KMS -> bbb-webrtc-sfu -> client.

An adjustment that can be done to improve scenarios like this are:
  - Bump https://github.com/bigbluebutton/bigbluebutton/blob/v2.2.x-release/bigbluebutton-html5/private/config/settings.yml#L107 to something bigger to give headroom (eg 30s)

**Comment by basisbit** [comment]
> thread starvation

is there any way I could test if this is actually the problem? We are running BBB on a Ryzen 3900 system with 12 cores / 24 cpu threads, I'd be surprised if you can cause thread starvation with 50 people enabling their webcam sharing within a short time frame on an otherwise empty bbb server - that would mean that the threads do not take other tasks while waiting x seconds for new messages on the websocket connection where a client is requesting/starting webcam sharing...

**Comment by prlanzarin** [comment]
No, I don't have any specific testing steps besides using callgrind which is not doable on a scenario like this. And I'm also not sure starvation is the problem (see _might_).

Deeper investigation would be needed to find which is the bottleneck. I've given my generic overview of possible causes. More probable causes are context switching overhead (just do a test and monitor the number of threads spawned by KMS/gstreamer and you'll get the idea) and a badly implemented outbound event system in KMS (which I mentioned in the last comment, about the flow in/out states). Both of them are equally possible. The event system was altered a while ago in KMS to prevent a frequent enough crash caused by a race condition and that involved locking sigc++ (the one that handles all outbound events and C <-> C++ events) calls around a recursive mutex.

And, also, be aware that the client does not directly communicate with Kurento. So comparing nginx usage and worker limits based with KMS usage is not completetly correct because Kurento's signalling plane does not use nginx.

The client talks with bbb-webrtc-sfu, which talks with Kurento. The process in webrtc-sfu which talks to clients is a different one than the one that talks to KMS. Only one WS connection is established with Kurento by the control plane process. 
You could say that since Node.js is JS-single threaded, it could be processing webcams "sequentially". But that's not entirely correct. I/O is concurrent with a pool of ie 8 threads and every method call in bbb-webrtc-sfu is non-blocking (which goes to the event loop, which, if uses I/O, is concurrent).

I'd like to see data on how you think that calls are not being done concurrently.
And I'd also like to know if you can share Kurento's logs.

**Comment by prlanzarin** [comment]
Another interesting piece of data that would be good to look at is a more concrete usage graph of a reproducible scenario. You could use, for instance, a dstat dump with full fledged metrics and plot it into a graph. See the script down below.

 ```
#!/bin/bash

DATE=`date +'%Y%m%d%H%M'`
dstat -tcmdngyi --noheaders --tcp --udp --output dstat-$DATE.out
```

**Comment by prlanzarin** [comment]
Another test I'd like to know the output of is running the thing you did with 3 Kurentos and measuring whether joining listen only or screen sharing churns out errors or have abnormally degraded negotiation times during **peak usage** (when cameras starts 
1020-oing).

Finally, please output WebRtcEndpoint.conf.ini with redacted values if necessary.

**Comment by basisbit** [comment]
[kurento video process.log](https://github.com/bigbluebutton/bigbluebutton/files/5781215/2021-01-02T061701.00000.pid13082.-.cut.-.cleaned.log)

**Comment by basisbit** [comment]
WebRtcEndpoint.conf.ini:
```
pemCertificateRSA=/etc/kurento/dtls-srtp.pem
externalIPv4=123.123.123.123
```

**Comment by prlanzarin** [comment]
I'll take a look as soon as I can. Thanks.

**Comment by hbinkle** [comment]
is this issue somehow related to any client cookies?
We noticed that once we got error 1020 and the black screen we are no longer able to activate the cam, even hours or days later.
Rejoining the session using an incognito-tab succeeds and the cam is back.

**Comment by basisbit** [comment]
@hbinkle no, this issue here is not related to cookies. I'd suggest you to ask at the bigbluebutton-setup mailing list regarding your issue - probably kurento died on the server and eventually was restarted automagically by the cron job.

**Comment by flowerpower** [comment]
I have the same issue on a BBB cluster (5 nodes with **BBB 2.2.31**), happened 2 times in different days on two differents BBB nodes. Happened with 30 - 50 concurrent webcam activations in a short time (students connection for a lesson).
Have you solved the problem?
And #11120 can help?
Thank you in advance

**Comment by basisbit** [comment]
@flowerpower please read the messages above, they already answer most of your questions. Also read the description of the PR which you linked to.

> Have you solved the problem?

This issue here has the status `open`, so no, it is not yet closed.

**Comment by prlanzarin** [comment]
#11120 was suggested as a out of the box mitigation.
If you want another thing that can help, enable [video pagination](https://github.com/bigbluebutton/bigbluebutton/pull/10293).
If you put a feasible page size, it should reduce the occurrence of this scenario to a bare minimum because it constrains complexity

I'm looking at this at my free time to see if I can find room for improving the negotiation phase CPU usage and I'll update this issue with info when (and if) I get something out of it (not before, not after).

Moreover, don't expect wonders. I've outlined the structural problems that cause this and it's no small deal.

**Comment by flowerpower** [comment]
> 
> 
> @flowerpower please read the messages above, they already answer most of your questions. Also read the description of the PR which you linked to.
> 
> > Have you solved the problem?
> 
> This issue here has the status `open`, so no, it is not yet closed.

Im' sorry for the question, but the bug is opened 27 days ago and I thinked that was an abandoned issue

**Comment by flowerpower** [comment]
> 
> 
> #11120 was suggested as an out of the box mitigation of low impact.
> If you want another thing that can help, enable [video pagination](https://github.com/bigbluebutton/bigbluebutton/pull/10293).
> If you put a feasible page size, it should reduce the occurrence of this scenario to a bare minimum because it constrains complexity
> 
> I'm looking at this at my free time to see if I can find room for improving the negotiation phase CPU usage and I'll update this issue with info when (and if) I get something out of it (not before, not after).
> 
> Moreover, don't expect wonders. I've outlined the structural problems that cause this and it's no small deal.

The problem is happend for me on a phisical server with 64 CPU and 128GB RAM, with ~40 concurrent webcam activations.
Thanks for your time

**Comment by basisbit** [comment]
by the way, the system that we use is running on decent hardware, so using bare metal over VMs does not "solve" the issue.
Imho, this is a problem where we probably need some Kurento developer. Maybe someone could get commercial support from Kurento Technologies?

**Comment by prlanzarin** [comment]
Just some stuff I commented in the mailing lists regarding this issue that I think should be registered here:

> https://gstreamer.freedesktop.org/data/events/gstreamer-conference/2018/Sebastian%20Dr%C3%B6ge%20-%20When%20adding%20more%20threads%20adds%20more%20problems:%20Thread-sharing%20between%20elements%20in%20GStreamer.pdf

> There are theoretical approaches to alleviate the problematic scenarios (> 1k streams), but we've not implemented (or finished) any of them yet. For example:
>   - Negotiation priority queues based on media directions (ie prioritize negotiating sendonly streams, aka shared cameras, over recvonly streams, aka subscribers)
>   - Negotiation pacing/throttling: batch negotiation in queues of N streams based on the current number of streams in the meeting (avoid spiralling)
>   - https://github.com/bigbluebutton/bbb-webrtc-sfu/issues/65


> A properly virtualized, dedicated VM from a reliable provider (Digital Ocean, AWS, Hetzner, Oracle, ...) should be as good as a bare metal instance in regards
> to that specific problem.

> Actual mitigations for that issue are:
>   - increasing the base media timeouts (suggested in the issue)
>   - enabling video pagination (suggested in the issue)
>   - enabling dynamic video profiles (cameraQualityThresholds in settings.yml)
>   - use the networkInterfaces config option in Kurento. It's configured in /etc/kurento/modules/kurento/WebRtcEndpoint.conf.ini. You should set
> it to whichever is the default, public network interface in your server (in this case, ens2).
>     * If the NIF has unwanted IPs (ie private intranet IPs for VPS comms which are not used for media exchange), you can add them to ipIgnoreList
>     * It's even a bit hard to stress how important this config is; please set it. It speeds up connectivity checks and reduces signalling overhead a LOT
>  - set niceAgentIceTcp to 0 (/etc/kurento/modules/kurento/WebRtcEndpoint.conf.ini) if you're not using TCP for media comm (that's BBB's default, so
> it can be set to 0)
>  - /etc/kurento/modules/kurento/WebRtcEndpoint.conf.ini -> Prefer externalIPv4 if your server is behind NAT over stunServerAddress/Port
>  - if you're on Scaleway or some other provider that might be finicky with DNS, be sure to do the following:
> 
>     change kurentoAllowMDNSCandidates to false in /usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml
>     sudo systemctl restart bbb-webrtc-sfu


> Couple Kurento's bad vertical scaling with the bad time browsers have when running multiple peer connections
> (one per track, BBB's current approach), you'll either have Kurento taking a bit to negotiate stuff or the client's
> machines struggling a lot (both scenarios render 1020s).

**Comment by prlanzarin** [comment]
> I'm looking at this at my free time to see if I can find room for improving the negotiation phase CPU usage and I'll update this issue with info when (and if) I get something out of it (not before, not after).

Also: I'm not doing this (negotiation phase CPU/context switching efficiency), specifically, anymore.

I'm looking to improve this in a wider spectrum.

**Comment by robregonm** [comment]
I experienced the same issue in a meeting with about 100 students, but only a few cameras were enabled (5~10 cams).
Most of students were unable to see the cameras, they just saw black screens.
I tried in a 32 CPU servers, 32~64 GB RAM and BBB version 2.2.34.
CPU usage was about 30~50%, and there was plenty of RAM available

**Comment by prlanzarin** [comment]
Read all the mitigations that were suggested in this thread, try applying them.

If the behaviour is still the same:
  - CPU and network usage data
  - Logs (if possible) (/var/log/kurento-media-server/, /var/log/bbb-webrtc-sfu.log)
  
> This is kinda weird behaviour because in the same servers we had 2x-3x more students (different meetings) and many more cameras simultaneously without issues. The only difference is that in this case, all these students were in a single meeting.

That's not necessarily weird. Having 100 meetings with 3 users and three webcams each (total 300 inbound cameras), for instance, == 900 streams. Having a single meeting with 100 users and 10 webcams == 1000 streams.
Stream complexity is quadratic.

**Comment by prlanzarin** [comment]
As the focus here was Kurento-related bottlenecks, I'll close this as addressed in 2.5 (default adapter is mediasoup, etc).
